batch/batch/worker/copy_files.py

```diff
@@ -0,0 +1,342 @@
+    if file_size == 0:
+        return ([0, 0], 1)
+
+    n_partitions = (file_size + min_partition_size - 1) // min_partition_size
+
+    if max_partitions:
+        n_partitions = min(max_partitions, n_partitions)
+
+    partition_size = file_size // n_partitions
+    n_partitions_w_extra_byte = file_size % n_partitions
+
+    partition_sizes = [partition_size + 1] * n_partitions_w_extra_byte + \
+                      [partition_size] * (n_partitions - n_partitions_w_extra_byte)
+
+    pos = 0
+    boundaries = [0]
+    for size in partition_sizes:
+        pos += size
+        boundaries.append(pos)
+
+    return (boundaries, n_partitions)
+
+
+async def glob_gcs(path):
+    return [('gs://' + blob.bucket.name + '/' + blob.name, blob.size)
+            for blob in await gcs_client.glob_gs_files(path)]
+
+
+async def glob_local(path):
+    return await async_os.glob(path)
+
+
+async def is_gcs_dir(path):
+    return is_gcs_path(path) and \
+           (path.endswith('/') or len(await gcs_client.glob_gs_files(path.rstrip('/') + '/*')) != 0)  # FIXME: Add max_results to glob
+
+
+async def is_local_dir(path):
+    # gsutil does not copy empty local directories
+    return not is_gcs_path(path) and \
+           (path.endswith('/') or len(await async_os.glob(path.rstrip('/') + '/*')) != 0)
+
+
+async def is_dir(path):
+    if is_gcs_path(path):
+        return await is_gcs_dir(path)
+    return await is_local_dir(path)
+
+
+async def dir_exists(path):
+    if is_gcs_path(path):
+        return next(await gcs_client.list_gs_files(path + '/', max_results=1), None) is not None
+    return os.path.isdir(path)
+
+
+def any_recursive_match(pattern, paths):
+    pattern = pattern.split('/')
+    x = any([len(pattern) != len(path.split('/')) for path in paths])
+    return x
+
+
+async def add_destination_paths(src, glob_paths, dest):
+    if not is_gcs_path(dest):
+        has_recursive_match = any_recursive_match(src, [p for p, _ in glob_paths])
+        if not has_recursive_match and dest.endswith('/') and not await dir_exists(dest):
+            if len(glob_paths) == 1:
+                if is_gcs_path(src):
+                    print('skipping destination file ending with slash')
+                    return []
+                print('destination is a directory')
+                raise IsADirectoryError
+            if len(glob_paths) > 1:
+                print('destination must name a directory when matching multiple files')
+                raise NotADirectoryError
+
+    is_dir_src = await is_dir(src)
+    is_dir_dest = await is_dir(dest)
+
+    if not is_dir_src and not is_dir_dest:
+        if len(glob_paths) == 0:
+            raise FileNotFoundError
+        if len(glob_paths) == 1:
+            return [(file, dest, size) for file, size in glob_paths]
+        if is_gcs_path(dest):
+            include_recurse_dir = not is_gcs_path(dest) or await is_gcs_dir(dest)  # DOUBLE CHECK THIS!!!
```

Comment:
`is_gcs_path(dest)` must be `True` and `is_gcs_dir(dest)` must be `False`, so we can just set this to `False`